src/main/kotlin/ecommerce/interceptor/AuthInterceptor.kt

```diff
@@ -0,0 +1,106 @@
+package ecommerce.interceptor
+
+import ecommerce.service.TokenService
+import jakarta.servlet.http.HttpServletRequest
+import jakarta.servlet.http.HttpServletResponse
+import org.springframework.stereotype.Component
+import org.springframework.web.servlet.HandlerInterceptor
+
+@Component
+class AuthInterceptor(private val tokenService: TokenService) : HandlerInterceptor {
+    companion object {
+        const val USER_ID_ATTRIBUTE = "userId"
+        const val USER_ROLE_ATTRIBUTE = "userRole"
+        const val USER_EMAIL_ATTRIBUTE = "userEmail"
+        const val USER_NAME_ATTRIBUTE = "userName"
+        private const val BEARER_PREFIX = "Bearer "
+        private const val ADMIN_PATH_PREFIX = "/admin/"
+        private const val ADMIN_ROLE = "ADMIN"
+    }
+
+    override fun preHandle(
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        handler: Any,
+    ): Boolean {
+        return try {
+            val token =
+                extractToken(request) ?: return sendUnauthorizedResponse(
+                    response, "Missing or invalid Authorization header",
+                )
+
+            val claims =
+                tokenService.validateToken(token) ?: return sendUnauthorizedResponse(
+                    response, "Invalid or expired token",
+                )
+
+            val userId =
+                extractUserId(claims) ?: return sendUnauthorizedResponse(
+                    response, "Invalid token payload",
+                )
+
+            storeUserAttributes(request, claims, userId)
+
+            if (requiresAdminAccess(request) && !hasAdminRole(claims)) {
+                return sendForbiddenResponse(response, "Admin access required")
+            }
+
+            true
+        } catch (e: Exception) {
+            sendUnauthorizedResponse(response, "Authentication error: ${e.message}")
+        }
+    }
+
+    private fun extractToken(request: HttpServletRequest): String? {
+        val authHeader = request.getHeader("Authorization")
+        return if (authHeader?.startsWith(BEARER_PREFIX) == true) {
+            authHeader.substring(BEARER_PREFIX.length)
+        } else {
+            null
+        }
+    }
+
+    private fun extractUserId(claims: io.jsonwebtoken.Claims): Long? {
+        return claims.subject?.toLongOrNull()
+    }
+
+    private fun storeUserAttributes(
+        request: HttpServletRequest,
+        claims: io.jsonwebtoken.Claims,
+        userId: Long,
+    ) {
+        request.setAttribute(USER_ID_ATTRIBUTE, userId)
+        request.setAttribute(USER_ROLE_ATTRIBUTE, claims["role"] as? String)
+        request.setAttribute(USER_EMAIL_ATTRIBUTE, claims["email"] as? String)
+        request.setAttribute(USER_NAME_ATTRIBUTE, claims["name"] as? String)
```

Comment:
Instead of storing multiple string fields separately, how about encapsulating them in one object and storing it together? That would improve cohesion. 🙂